activerecord/lib/active_record/connection_adapters.rb

```diff
@@ -4,6 +4,65 @@ module ActiveRecord
   module ConnectionAdapters
     extend ActiveSupport::Autoload
 
+    @adapters = {}
+
+    class << self
+      # Registers a custom database adapter.
+      #
+      # Can also be used to define aliases.
+      #
+      # == Example
+      #
+      #   ActiveRecord::ConnectionAdapters.register("megadb", "MegaDB::ActiveRecordAdapter", "mega_db/active_record_adapter")
+      #
+      #   ActiveRecord::ConnectionAdapters.register("mysql", "ActiveRecord::ConnectionAdapters::TrilogyAdapter", "active_record/connection_adapters/trilogy_adapter")
+      #
+      def register(name, class_name, path = class_name.underscore)
+        @adapters[name] = [class_name, path]
+      end
+
+      def resolve(adapter_name) # :nodoc:
+        # Require the adapter itself and give useful feedback about
+        #   1. Missing adapter gems and
+        #   2. Adapter gems' missing dependencies.
+        class_name, path_to_adapter = @adapters[adapter_name]
+
+        unless class_name
+          raise AdapterNotFound, "database configuration specifies nonexistent '#{adapter_name}' adapter. Ensure that the adapter is spelled correctly in config/database.yml and that you've added the necessary adapter gem to your Gemfile."
+        end
+
+        unless Object.const_defined?(class_name)
+          begin
+            require path_to_adapter
+          rescue LoadError => error
+            # We couldn't require the adapter itself. Raise an exception that
+            # points out config typos and missing gems.
```

Comment:
Thought I'd give this a go, but I'm questioning the usefulness of this in the error you suggested if my understanding below is correct.

If we're already at the point of requiring the path, doesn't that mean that the gem has probably already been added and therefore the adapter has been registered? Else `AdapterNotFound` would have already been thrown before this. Meaning the first `LoadError` is only due to an incorrect registered path by the custom adapter gem. I'm failing to understand how this could be a user error / typo and how a did-you-mean would help if this is case.

Could we maybe upgrade the `AdapterNotFound` with the hint instead? In that case it's possible that the `adapter_name` was a typo and we may already have a registered adapter to use for the did-you-mean.

Reply:
Ah, yes. Perhaps the below error could now be more explicit that it's describing an implementation error, rather that app-level configuration. (It _used_ to be the main user-facing "typo maybe?" error, but as you note, now is not.)

Reply:
(but my general did-you-mean suggestion does apply to the earlier AdapterNotFound :+1:)

Reply:
Got this up: https://github.com/rails/rails/pull/50104